src/marks/auto.js

```diff
@@ -0,0 +1,328 @@
+import {isOrdinal, labelof, valueof, isOptions} from "../options.js";
+import {area, areaX, areaY} from "./area.js";
+import {dot} from "./dot.js";
+import {line, lineX, lineY} from "./line.js";
+import {ruleX, ruleY} from "./rule.js";
+import {barX, barY} from "./bar.js";
+import {rect, rectX, rectY} from "./rect.js";
+import {cell} from "./cell.js";
+import {frame} from "./frame.js";
+import {bin, binX, binY} from "../transforms/bin.js";
+import {group, groupX, groupY} from "../transforms/group.js";
+import {marks} from "../plot.js";
+import {ascending} from "d3";
+
+// https://github.com/observablehq/stdlib/blob/main/src/table.js
+const nChecks = 20; // number of values to check in each array
+
+export function auto(data, {x, y, fx, fy, color, size, mark} = {}) {
+  // Shorthand: array of primitives should result in a histogram
+  if (x === undefined && y === undefined && arrayIsPrimitive(data)) x = (d) => d;
+
+  // Allow x and y and other dimensions to be specified as shorthand
+  // field names (but note that they can also be specified as a
+  // {transform} object such as Plot.identity).
+  if (!isOptions(x)) x = makeOptions(x);
+  if (!isOptions(y)) y = makeOptions(y);
+  if (!isOptions(fx)) fx = makeOptions(fx);
+  if (!isOptions(fy)) fy = makeOptions(fy);
+  if (!isOptions(color)) color = makeOptions(color);
+  if (!isOptions(size)) size = makeOptions(size);
+
+  const {value: xValue} = x;
+  const {value: yValue} = y;
+
+  const {value: sizeValue, reduce: sizeReduce} = size;
+
+  // Determine the default reducer, if any.
+  let {reduce: xReduce} = x;
+  let {reduce: yReduce} = y;
+  if (xReduce === undefined)
+    xReduce = yReduce == null && xValue == null && sizeValue == null && yValue != null ? "count" : null;
+  if (yReduce === undefined)
+    yReduce = xReduce == null && yValue == null && sizeValue == null && xValue != null ? "count" : null;
+
+  const {zero: xZero} = x;
+  const {zero: yZero} = y;
+  const {value: colorValue, reduce: colorReduce} = color;
```

Comment:
Let’s support this for unambiguously specifying a constant color:

```js
Plot.auto(penguins, {x: "culmen_length_mm", color: {color: "red"}})
```

And then the shorthand string color needs to disambiguate. The above is the same as this:

```js
Plot.auto(penguins, {x: "culmen_length_mm", color: "red"})
```

Whereas

```js
Plot.auto(penguins, {x: "culmen_length_mm", color: "foo"})
```

would be shorthand for

```js
Plot.auto(penguins, {x: "culmen_length_mm", color: {value: "foo"}})
```